File: VB/ShowDetailViewFromDashboard.Module.Web/Controllers/WebShowDetailViewFromDashboardController.vb
Imports DevExpress.DashboardWeb
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Dashboards.Web
Imports DevExpress.ExpressApp.Web
Imports DevExpress.ExpressApp.Web.Templates
Imports DevExpress.Persistent.Base
Imports ShowDetailViewFromDashboard.Module.BusinessObjects
Imports System

Namespace ShowDetailViewFromDashboard.Module.Web.Controllers

    Public Class WebShowDetailViewFromDashboardController
        Inherits ObjectViewController(Of DetailView, IDashboardData)
        Implements IXafCallbackHandler

        Private Const HandlerName As String = "WebShowDetailViewFromDashboardController"

        Private ReadOnly openDetailViewAction As ParametrizedAction

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Dim dashboardViewerViewItem As WebDashboardViewerViewItem = TryCast(View.FindItem("DashboardViewer"), WebDashboardViewerViewItem)
            If dashboardViewerViewItem IsNot Nothing Then
                If dashboardViewerViewItem.DashboardDesigner IsNot Nothing Then
                    CustomizeDashboardViewer(dashboardViewerViewItem.DashboardDesigner)
                End If

                dashboardViewerViewItem.ControlCreated += AddressOf DashboardViewerViewItem_ControlCreated
            End If
        End Sub

        Private Sub DashboardViewerViewItem_ControlCreated(ByVal sender As Object, ByVal e As EventArgs)
            Dim dashboardViewerViewItem As WebDashboardViewerViewItem = TryCast(sender, WebDashboardViewerViewItem)
            CustomizeDashboardViewer(dashboardViewerViewItem.DashboardDesigner)
        End Sub

        Private Sub CustomizeDashboardViewer(ByVal dashboardControl As ASPxDashboard)
            Dim callbackManager As XafCallbackManager = CType(WebWindow.CurrentRequestPage, ICallbackManagerHolder).CallbackManager
            callbackManager.RegisterHandler(HandlerName, Me)
            Dim widgetScript As String = "function getOid(s, e) {{" & Microsoft.VisualBasic.Constants.vbCrLf & "                                        function findMeasure(measure) {{" & Microsoft.VisualBasic.Constants.vbCrLf & "                                            return measure.DataMember === 'Oid';" & Microsoft.VisualBasic.Constants.vbCrLf & "                                        }}" & Microsoft.VisualBasic.Constants.vbCrLf & "                                        if (e.ItemName.includes('gridDashboardItem')) {{" & Microsoft.VisualBasic.Constants.vbCrLf & "                                             var axisPoint = e.GetAxisPoint();" & Microsoft.VisualBasic.Constants.vbCrLf & "                                             if (!axisPoint) " & Microsoft.VisualBasic.Constants.vbCrLf & "                                                 return null;" & Microsoft.VisualBasic.Constants.vbCrLf & "                                             var itemData = e.GetData(),   " & Microsoft.VisualBasic.Constants.vbCrLf & "                                                dataSlice = itemData.GetSlice(axisPoint)," & Microsoft.VisualBasic.Constants.vbCrLf & "                                                oidMeasure = dataSlice.GetMeasures().find(findMeasure).Id," & Microsoft.VisualBasic.Constants.vbCrLf & "                                                measureValue = dataSlice.GetMeasureValue(oidMeasure);" & Microsoft.VisualBasic.Constants.vbCrLf & "                                                {0}" & Microsoft.VisualBasic.Constants.vbCrLf & "                                        }}" & Microsoft.VisualBasic.Constants.vbCrLf & "                                    }}"
            dashboardControl.ClientSideEvents.ItemClick = String.Format(widgetScript, callbackManager.GetScript(HandlerName, "measureValue.GetValue()"))
        End Sub

        Public Sub ProcessAction(ByVal parameter As String)
            openDetailViewAction.DoExecute(parameter)
        End Sub

        Private Sub OpenDetailViewAction_Execute(ByVal sender As Object, ByVal e As ParametrizedActionExecuteEventArgs)
            Dim oid As Guid = Nothing
            Guid.TryParse(CStr(e.ParameterCurrentValue), oid)
            Dim objectSpace As IObjectSpace = Application.CreateObjectSpace(GetType([Module].BusinessObjects.Contact))
            Dim contact As [Module].BusinessObjects.Contact = objectSpace.FindObject(Of [Module].BusinessObjects.Contact)(New BinaryOperator(NameOf([Module].BusinessObjects.Contact.Oid), oid))
            If contact IsNot Nothing Then
                e.ShowViewParameters.CreatedView = Application.CreateDetailView(objectSpace, contact, View)
            End If
        End Sub

        Protected Overrides Sub OnDeactivated()
            Dim dashboardViewerViewItem As WebDashboardViewerViewItem = TryCast(View.FindItem("DashboardViewer"), WebDashboardViewerViewItem)
            If dashboardViewerViewItem IsNot Nothing Then
                dashboardViewerViewItem.ControlCreated -= AddressOf DashboardViewerViewItem_ControlCreated
            End If

            MyBase.OnDeactivated()
        End Sub

        Public Sub New()
            openDetailViewAction = New ParametrizedAction(Me, "Dashboard_OpenDetailView", "Dashboard", GetType(String))
            openDetailViewAction.Caption = "OpenDetailView"
            openDetailViewAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject
            Me.openDetailViewAction.Execute += AddressOf OpenDetailViewAction_Execute
        End Sub
    End Class
End Namespace

File: CS/ShowDetailViewFromDashboard.Module.Web/Controllers/WebShowDetailViewFromDashboardController.cs
using DevExpress.DashboardWeb;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Dashboards.Web;
using DevExpress.ExpressApp.Web;
using DevExpress.ExpressApp.Web.Templates;
using DevExpress.Persistent.Base;
using ShowDetailViewFromDashboard.Module.BusinessObjects;
using System;

namespace ShowDetailViewFromDashboard.Module.Web.Controllers {
    public class WebShowDetailViewFromDashboardController : ObjectViewController<DetailView, IDashboardData>, IXafCallbackHandler {
        private const string HandlerName = "WebShowDetailViewFromDashboardController";
        private readonly ParametrizedAction openDetailViewAction;

        protected override void OnActivated() {
            base.OnActivated();
            WebDashboardViewerViewItem dashboardViewerViewItem = View.FindItem("DashboardViewer") as WebDashboardViewerViewItem;
            if (dashboardViewerViewItem != null) {
                if (dashboardViewerViewItem.DashboardDesigner != null) {
                    CustomizeDashboardViewer(dashboardViewerViewItem.DashboardDesigner);
                }
                dashboardViewerViewItem.ControlCreated += DashboardViewerViewItem_ControlCreated;
            }
        }
        private void DashboardViewerViewItem_ControlCreated(object sender, EventArgs e) {
            WebDashboardViewerViewItem dashboardViewerViewItem = sender as WebDashboardViewerViewItem;
            CustomizeDashboardViewer(dashboardViewerViewItem.DashboardDesigner);
        }
        private void CustomizeDashboardViewer(ASPxDashboard dashboardControl) {
            XafCallbackManager callbackManager = ((ICallbackManagerHolder)WebWindow.CurrentRequestPage).CallbackManager;
            callbackManager.RegisterHandler(HandlerName, this);
            string widgetScript = @"function getOid(s, e) {{
                                        function findMeasure(measure) {{
                                            return measure.DataMember === 'Oid';
                                        }}
                                        if (e.ItemName.includes('gridDashboardItem')) {{
                                             var axisPoint = e.GetAxisPoint();
                                             if (!axisPoint) 
                                                 return null;
                                             var itemData = e.GetData(),   
                                                dataSlice = itemData.GetSlice(axisPoint),
                                                oidMeasure = dataSlice.GetMeasures().find(findMeasure).Id,
                                                measureValue = dataSlice.GetMeasureValue(oidMeasure);
                                                {0}
                                        }}
                                    }}";
            dashboardControl.ClientSideEvents.ItemClick = string.Format(widgetScript, callbackManager.GetScript(HandlerName, "measureValue.GetValue()"));
        }
        public void ProcessAction(string parameter) {
            openDetailViewAction.DoExecute(parameter);
        }
        private void OpenDetailViewAction_Execute(object sender, ParametrizedActionExecuteEventArgs e) {
            Guid.TryParse((string)e.ParameterCurrentValue, out var oid); 
            IObjectSpace objectSpace = Application.CreateObjectSpace(typeof(Contact));
            Contact contact = objectSpace.FindObject<Contact>(new BinaryOperator(nameof(Contact.Oid), oid));
            if (contact != null) {
                e.ShowViewParameters.CreatedView = Application.CreateDetailView(objectSpace, contact, View);
            }
        }

        protected override void OnDeactivated() {
            WebDashboardViewerViewItem dashboardViewerViewItem = View.FindItem("DashboardViewer") as WebDashboardViewerViewItem;
            if (dashboardViewerViewItem != null) {
                dashboardViewerViewItem.ControlCreated -= DashboardViewerViewItem_ControlCreated;
            }
            base.OnDeactivated();
        }
        public WebShowDetailViewFromDashboardController() {
            openDetailViewAction = new ParametrizedAction(this, "Dashboard_OpenDetailView", "Dashboard", typeof(string));
            openDetailViewAction.Caption = "OpenDetailView";
            openDetailViewAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject;
            openDetailViewAction.Execute += OpenDetailViewAction_Execute;
        }
    }
}
